crates/binstalk-downloader/src/gh_api_client.rs

```diff
@@ -0,0 +1,398 @@
+use std::{
+    collections::{HashMap, HashSet},
+    ops::Deref,
+    sync::{Arc, Mutex, RwLock},
+    time::{Duration, Instant},
+};
+
+use compact_str::{CompactString, ToCompactString};
+use tokio::sync::OnceCell;
+
+use crate::remote;
+
+mod request;
+pub use request::GhApiError;
+
+/// default retry duration if x-ratelimit-reset is not found in response header
+const DEFAULT_RETRY_DURATION: Duration = Duration::from_secs(3);
+
+/// The keys required to identify a github release.
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct GhRelease {
+    pub owner: CompactString,
+    pub repo: CompactString,
+    pub tag: CompactString,
+}
+
+/// The Github Release and one of its artifact.
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct GhReleaseArtifact {
+    pub release: GhRelease,
+    pub artifact_name: CompactString,
+}
+
+impl GhReleaseArtifact {
+    /// Create [`GhReleaseArtifact`] from url.
+    pub fn try_extract_from_url(url: &remote::Url) -> Option<Self> {
+        if url.domain() != Some("github.com") {
+            return None;
+        }
+
+        let mut path_segments = url.path_segments()?;
+
+        let owner = path_segments.next()?;
+        let repo = path_segments.next()?;
+
+        if (path_segments.next()?, path_segments.next()?) != ("releases", "download") {
+            return None;
+        }
+
+        let tag = path_segments.next()?;
+        let artifact_name = path_segments.next()?;
+
+        (path_segments.next().is_none() && url.fragment().is_none() && url.query().is_none()).then(
+            || Self {
+                release: GhRelease {
+                    owner: owner.to_compact_string(),
+                    repo: repo.to_compact_string(),
+                    tag: tag.to_compact_string(),
+                },
+                artifact_name: artifact_name.to_compact_string(),
+            },
+        )
+    }
+}
+
+#[derive(Debug)]
+struct Map<K, V>(RwLock<HashMap<K, Arc<V>>>);
```

Comment:
This seems like it would have pretty bad performance patterns.

To be clear, it's a:

```rust
RwLock<HashMap<GhRelease, Arc<OnceCell<Option<HashSet<CompactString>>>>>>
```

With `Lock<Map<_, Arc<OnceCell<Option<Set` we're doing the equivalent of:

```rust
// read lock open
if let Some(r) = artifacts.read().get(release) {
	return r;
}
// read lock closed

// write lock open
let handle = artifacts.write().entry(release);

// fetch artifacts

handle.or_insert(Arc::new(names));
drop(handle);
// write lock closed
```

ie we're always getting a write lock for the duration of the artifact grab future if the value isn't present, blocking all other readers.

Why not have a `Lock<Map<_, Arc<Set` (`Option<Set` expressed as an empty `Set` instead of `None`) and doing something like:

```rust
// read lock open
if artifacts.read().has(release) {
  return;
}
// read lock closed

// get the artifacts

// write lock open
artifacts.write().entry(release).or_insert(Arc::new(names))
// write lock closed
```

or using a concurrent map of some sort that can do multiple writers.

Reply:
I used `OnceCell` to ensure that they would only submit one restful API request for the entire GitHub Release because the rate limit for unauthorized users are really tight (60 per hour per originating IP address).

For `Option<HashSet<...>>`, `None` represents no such GitHub Release and `Some(...)` represents that there is indeed a GitHub Release.

It's possible to have a GitHub Release with no artifacts so I don't think using empty `HashSet<...>` to represent this could work.

> or using a concurrent map of some sort that can do multiple writers.

IMO we probably don't need something that advanced, the existing solution already enables concurrent access.

I've checked other concurrent map before and often they use multiple shard (`HashMap<...>`) internally or something that assumes read-heavy, use two `HashMap<...>`, one for read and one for write plus a write log, and requires a commit for the write to be seen (joohoo's left-right crate).

Reply:
BTW, I'm also considering using GitHub Restful API to speedup `QuickInstall`, which encodes the version and target all in release/tag name.

Unfortunately there is no search API for the release or tag, so I'm considering [listing release tags](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags) since listing release would contains too many unused information and there are many, many tags on cargo-quickinstall.

But even then, we cannot guarantee any tag of interest is necessary cached so not sure how to do this efficiently.

Doing so probably requires cooperation from cargo-quickinstall, like storing a cache for all artifacts for a crate or a specific version ranges of a crate (e.g. cargo-binstall >=0.12.1,<=0.20.1).

cc @alsuren @Milo123459

Reply:
P.S. I removed an unrelated comment from this